src/common/utils/refct_ptr.h

```diff
@@ -0,0 +1,72 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#ifndef SRC_COMMON_UTILS_REFCT_PTR_H_
+#define SRC_COMMON_UTILS_REFCT_PTR_H_
+
+#include "src/common/utils/intrusive_ptr.h"
+#include "src/common/utils/ref_counted.h"
+
+namespace raksha::common::utils {
+
+template <class T>
+class WithRefCount : public RefCounted<WithRefCount<T>> {
+ public:
+  template <class... Args>
+  WithRefCount(Args&&... args)
+      : RefCounted<WithRefCount<T>>(), payload_(std::forward<Args>(args)...) {}
+
+  T& payload() { return payload_; }
+  const T& payload() const { return payload_; }
+
+ private:
+  T payload_;
+};
+
+// An intrusive_ptr that only intrudes on the `WithRefCount` object it
+// introduces. As such, it's not really "intrusive" from the point of view of
+// the client and deserves a new name.
+template <class T>
+class refct_ptr {
```

Comment:
Currently, we're looking at using `intrusive_ptr` in the Datalog IR structures to make them copyable and thus easier to work with in circumstances such as tests. Currently, the main field of a `Record` is a `unique_ptr` of a `std::tuple` of all of the types that are tracked by that `Record`. If we were to try to have this be a `RefCounted` target, I suppose we'd have to create a new `RefCountedTuple` that inherits from `RefCounted` but is otherwise a tuple.

Then, we have ADTs, which currently contain within them a vector of `unique_ptr`s to their arguments. Each of those could be any Souffle `Value`, so I think we'd have to create a `RefCountedValue`, maybe a `RefCounted` version for each `Value`.

I'm basically seeing an explosion of `RefCounted` variants of types that come about due to `intrusive_ptr`'s insistence that its targets inherit from `RefCounted`. That feels wrong. I think perhaps the `RefCounted`-ness should be done via composition rather than inheritence (as I am trying to do here) to prevent this growth of `RefCounted` subclasses that need to be created and managed.

The more I interact with `intrusive_ptr`, the more I wonder if we should just be using `std::shared_ptr`. Yes, it is discouraged by the Google style guide, but only because the Google style guide encourages objects to have one canonical owner when possible - an issue that `intrusive_ptr` shares. In addition, `std::shared_ptr` is already written for us, no additional thinking or testing required.

Reply:
I think it is worth having a face to face discussion on this topic. Let us chat more when you get back and we can summarize our discussion here.